src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3TimeoutTests.cs

```diff
@@ -192,6 +196,376 @@ public async Task ControlStream_RequestHeadersTimeoutMaxValue_ExpirationIsMaxVal
+
+            // _helloWorldBytes is 12 bytes, and 12 bytes / 240 bytes/sec = .05 secs which is far below the grace period.
+            await requestStream.SendHeadersAsync(ReadRateRequestHeaders(_helloWorldBytes.Length), endStream: false);
+            await requestStream.SendDataAsync(_helloWorldBytes, endStream: false);
+
+            await requestStream.ExpectHeadersAsync();
+
+            await requestStream.ExpectDataAsync();
+
+            // Don't send any more data and advance just to and then past the grace period.
+            AdvanceClock(limits.MinRequestBodyDataRate.GracePeriod);
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(It.IsAny<TimeoutReason>()), Times.Never);
+
+            AdvanceClock(TimeSpan.FromTicks(1));
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(TimeoutReason.ReadDataRate), Times.Once);
+
+            await WaitForConnectionErrorAsync<ConnectionAbortedException>(
+                ignoreNonGoAwayFrames: false,
+                expectedLastStreamId: 8,
+                Http3ErrorCode.InternalError,
+                null);
+
+            _mockTimeoutHandler.VerifyNoOtherCalls();
+        }
+
+        /*
+         * Additional work around closing connections is required before response drain can be supported.
+        [Fact]
+        public async Task ResponseDrain_SlowerThanMinimumDataRate_AbortsConnection()
+        {
+            var mockSystemClock = _serviceContext.MockSystemClock;
+            var limits = _serviceContext.ServerOptions.Limits;
+
+            _timeoutControl.Initialize(mockSystemClock.UtcNow.Ticks);
+
+            await InitializeConnectionAsync(_noopApplication);
+
+            var inboundControlStream = await GetInboundControlStream();
+            await inboundControlStream.ExpectSettingsAsync();
+
+            CloseConnectionGracefully();
+
+            await inboundControlStream.ReceiveFrameAsync().DefaultTimeout();
+            await inboundControlStream.ReceiveFrameAsync().DefaultTimeout();
+            await inboundControlStream.ReceiveEndAsync().DefaultTimeout();
+
+            //await WaitForConnectionStopAsync(expectedLastStreamId: VariableLengthIntegerHelper.EightByteLimit, ignoreNonGoAwayFrames: false, expectedErrorCode: Http3ErrorCode.NoError);
+
+            AdvanceClock(TimeSpan.FromSeconds(inboundControlStream.BytesReceived / limits.MinResponseDataRate.BytesPerSecond) +
+                limits.MinResponseDataRate.GracePeriod + Heartbeat.Interval - TimeSpan.FromSeconds(.5));
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(It.IsAny<TimeoutReason>()), Times.Never);
+
+            AdvanceClock(TimeSpan.FromSeconds(1));
+
+            _mockTimeoutHandler.Verify(h => h.OnTimeout(TimeoutReason.WriteDataRate), Times.Once);
+
+            _mockTimeoutHandler.VerifyNoOtherCalls();
+        }
+        */
+
+        [Fact]
+        [QuarantinedTest("https://github.com/dotnet/aspnetcore/issues/21520")]
```

Comment:
Copied from HTTP/2. I have fixed its flakey issue here - https://github.com/dotnet/aspnetcore/pull/32139

I will update this PR with that fix once it is merged.